### PR TITLE
Improve dce

### PR DESCRIPTION
**Pull request by azidar**
Started from @terpstra's PR #490, but only the first commit (improvements to DCE).
Added tests a small bugfix.

### CODE REVIEW

File: src/main/scala/firrtl/passes/DeadCodeElimination.scala [hunk 1]
```diff
@@ -34,22 +34,22 @@ object DeadCodeElimination extends Pass {
       out
     }
 
-    def maybeEliminate(x: Statement, name: String) =
-      if (referenced(name)) x
+    def maybeEliminate(x: Statement, name: String, exception: String): Statement =
+      if (referenced(name) && name != exception) x
```
Comment (terpstra) [inline src/main/scala/firrtl/passes/DeadCodeElimination.scala]: Why is this necessary? The effectively neuters the DefRegister maybeEliminate. It now never happens.

**Comment by terpstra** [comment]
Is your intent with the 'exception' to detect cycles of logic that are otherwise unreferenced? If so, you should just use a mark and sweep algorithm after discovering edges.

**Comment by azidar** [comment]
No, it is just to catch the case where a register is initialized to itself. Otherwise, these registers will never be eliminated.

**Comment by terpstra** [comment]
Yes. That's exactly what I mean: a logic loop unreferenced anywhere else. If that is what you wanted to detect, then your algorithm is just a bad version of the mark-sweep algorithm, which is what we should use. The only reason I modified the pass as I did was that I wanted to keep my changes minimally invasive. However, the right way to do DCE if you want to detect cases like the one you are addressing is to do a complete mark-sweep pass. It will also be faster because it only needs to explore the netlist once.

**Comment by azidar** [comment]
In Firrtl, the way we describe an uninitialized register is by initializing it to itself. Thus, this specific use case is extremely common, and is very easy to check for. Otherwise, no uninitialized register will ever be DCE'd by this pass. This is also in the spirit of keeping things minimally invasive.

Perhaps "bugfix" was the wrong word, for that I apologize.

I agree with you that the more general case is a mark and sweep algorithm, for any logic loop, and this would catch this (as well as any other dead code this pass collects).

**Comment by jackkoenig** [comment]
Replaced by #549
